modules/is/manifests/init.pp

```diff
@@ -46,55 +32,72 @@
     system => true,
   }
 
-  # Ensure /opt/is directory is available
-  file { "/opt/${product}":
-    ensure => directory,
-    owner  => $user,
-    group  => $user_group,
+  # Create distribution path
+  file { [  "${products_dir}",
+            "${products_dir}/${product}" ]:
+    ensure  => 'directory',
+  }
+
+  # Change the ownership of the installation directory to wso2 user & group
+  file { $distribution_path:
+    ensure  => directory,
+    owner   => $user,
+    group   => $user_group,
+    require => [ User[$user], Group[$user_group]],
+    recurse => true
   }
 
-  # Copy the relevant installer to the /opt/is directory
-  file { "/opt/${product}/${product_package}":
+  # Copy binary to distribution path
+  file { "binary":
+    path   => "$distribution_path/${product_binary}",
     owner  => $user,
     group  => $user_group,
     mode   => '0644',
-    source => "puppet:///modules/${module_name}/${product_package}",
+    source => "puppet:///modules/${module_name}/${product_binary}",
   }
 
-  # Install WSO2 Identity Server
-  package { $package_name:
-    ensure   => installed,
-    provider => $installer_provider,
-    source  => "/opt/${product}/${product_package}"
+  # Stop the existing setup
+  exec { "stop-server":
+    command     => "kill -term $(cat ${install_path}/wso2carbon.pid)",
+    path        =>  "/bin/",
+    onlyif      => "/usr/bin/test -f ${install_path}/wso2carbon.pid",
+    subscribe   => File["binary"],
+    refreshonly => true,
   }
 
-  # Change the ownership of the installation directory to wso2 user & group
-  file { $install_path:
-    ensure  => directory,
-    owner   => $user,
-    group   => $user_group,
-    require => [ User[$user], Group[$user_group]],
-    recurse => true
+  # Wait for the server to stop
+  exec { "wait":
+    command     => "sleep 10",
+    path        =>  "/bin/",
+    onlyif      => "/usr/bin/test -d ${install_path}",
+    subscribe   => File["binary"],
+    refreshonly => true,
   }
 
-  # Copy configuration changes to the installed directory
-  $template_list.each |String $template| {
-    file { "${install_path}/${template}":
-      ensure  => file,
-      owner   => $user,
-      group   => $user_group,
-      mode    => '0644',
-      content => template("${module_name}/carbon-home/${template}.erb")
-    }
+  # Delete existing setup
+  exec { "delete-pack":
```

Comment:
Try to avoid "exec" as much as possible.
Use ensure => 'absent', recurse => 'true' to delete the directory.

Reply:
Exec is being used here instead of the file resource as the deletion of the deployment has to be dependent on (subscribed to) the "binary" file resource. If a File resource was used here instead of Exec, the agent will keep deleting the wso2is folder every 30 minutes. [File](https://puppet.com/docs/puppet/6.0/types/file.html) resource does not have a "Subscribe" attribute. So we have to use Exec in this case.